pkg/podrestartmonitor/pod_restart_monitor.go

```diff
@@ -0,0 +1,199 @@
+	"fmt"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"k8s.io/node-problem-detector/pkg/problemdaemon"
+	"k8s.io/node-problem-detector/pkg/types"
+)
+
+const PodRestartMonitorName = "pod-restart-monitor"
+
+func init() {
+	problemdaemon.Register(
+		PodRestartMonitorName,
+		types.ProblemDaemonHandler{
+			CreateProblemDaemonOrDie: NewMonitorOrDie,
+			CmdOptionDescription:     "Set to config file paths."})
+}
+
+type podRestartMonitor struct {
+	configPath string
+	config     MonitorConfig
+	stop       chan struct{}
+	clientset  *kubernetes.Clientset
+	statuses   chan *types.Status
+}
+
+// NewMonitorOrDie create a new podRestartMonitor, panic if error occurs.
+func NewMonitorOrDie(configPath string) types.Monitor {
+	l := &podRestartMonitor{
+		configPath: configPath,
+	}
+
+	f, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		glog.Fatalf("Failed to read configuration file %q: %v", configPath, err)
+	}
+	if l.config, err = parseConfig(f); err != nil {
+		glog.Fatalf("Failed to parse configuration file %q: %v", configPath, err)
+	}
+
+	glog.Infof("Finish parsing pod restart monitor config file %s: %+v", l.configPath, l.config)
+	l.stop = make(chan struct{})
+
+	return l
+}
+
+func parseConfig(f []byte) (config MonitorConfig, err error) {
+	if err = json.Unmarshal(f, &config); err != nil {
+		return config, err
+	}
+
+	// Apply default configurations
+	if config.CheckInterval == 0 {
+		config.CheckInterval = 5 * time.Minute
+	}
+	if config.RestartThreshold == 0 {
+		config.RestartThreshold = 5
+	}
+	if config.Namespace == "" {
+		config.Namespace = "kube-system"
+	}
+	if config.PodSelector == "" {
+		return config, fmt.Errorf("did not specify a PodSelector")
+	}
+	if config.ConditionName == "" {
+		config.ConditionName = ConditionTooManyRestarts
+	}
+	if _, err := labels.Parse(config.PodSelector); err != nil {
+		return config, fmt.Errorf("invalid PodSelector %q: %w", config.PodSelector, err)
+	}
+
+	return config, nil
+}
+
+func (l *podRestartMonitor) Start() (<-chan *types.Status, error) {
+	glog.Infof("Start pod restart monitor %s", l.configPath)
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	l.clientset, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	l.statuses = make(chan *types.Status)
+
+	go l.monitorLoop()
+
+	return l.statuses, nil
+}
+
+func getNodeName() string {
+	return os.Getenv("NODE_NAME")
+}
+
+func (l *podRestartMonitor) Stop() {
+	glog.Infof("Stopping monitor %s", l.configPath)
+	close(l.stop)
+}
+
+// ConditionTooManyRestarts indicates that a Pod has restarted too many times, indicating a problem with the node.
+const ConditionTooManyRestarts = "TooManyRestarts"
+
+// monitorLoop is the main loop of log monitor.
+func (l *podRestartMonitor) monitorLoop() {
```

Comment:
I think this comment, and the name `l`, are relics of this being copied from the log monitor package. It might be good to update these, or at least, the comment 😅